### DISCUSSION TITLE
java.lang.IllegalStateException: Client is currently closed. Must open messagingClient before sending

### DISCUSSION THREAD

**Discussion by vrmupparaju** [Azure/azure-iot-sdk-java#1743] [General]
Version Azure java sdk : 'com.microsoft.azure.sdk.iot:iot-service-client:2.1.6'

Below errors were throwing when I try to send messages to IOTHub

java.lang.IllegalStateException: Client is currently closed. Must open messagingClient before sending.

**below is the snippet**
try{
if(!messagingClient.isOpen()) {
messagingClient.open();
messageToSend.setDeliveryAcknowledgement(DeliveryAcknowledgement.None);
messagingClient.sendAsync(deviceId, messageToSend, null, null);

}catch(){
}
finally{
messagingClient.close();
}
I tried with send(deviceId,message) ,same issue persists.

Could somebody help me on this? or can share sample expamples

**Comment by timtay-microsoft** [comment]
You'll likely see this error if the client you opened loses its connection with the service. When that happens, you need to re-open the client to re-establish the connection. 

Here's some sample code that demonstrates how to do this:
https://github.com/Azure/azure-iot-service-sdk-java/blob/main/service/iot-service-samples/messaging-client-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/MessagingClientSample.java

**Reply by vrmupparaju** [reply]
@timtay-microsoft  Thank you so much for response.

I want to send a message to different devices thru IotHub, Added logic to open and close the connection for each call. Code is working fine with less devices. Below errors are occurring when I sent messages to no.of devices.
public void sendMessage(String deviceId,Message message){
messagingClient.open();
  messagingClient.send(deviceId, messageToSend); 
messagingClient.close();
}

**Errors:**
1. com.microsoft.azure.sdk.iot.service.exceptions.IotHubException: Message failed to send because the client was closed while it was still queued
2.com.microsoft.azure.sdk.iot.service.exceptions.IotHubException: Message failed to send because the client was closed after it was sent, but before it was acknowledged by the service.

Do I have to sleep the thread before closing connection for each call?

**Reply by timtay-microsoft** [reply]
When sending more than one message (to any number of devices) you should just leave the messaging client open until you are done with it.

I'm guessing you are calling that function from multiple places in parallel but using the same messagingClient instance within the function. As a result, one call to sendMessage may be in the middle of sending when another call to sendMessage is closing the client.
